File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        
        unordered_map<int,int>m;
        int l = 0;
        int res = 0;
        int si = m.size();
        for(int i = 0; i< s.size(); i++){

            while(m[s[i]] && l <= i){
                m[s[l]]--;
                if(!m[s[l]]){
                    m.erase(s[l]);
                }
                l++;
            }
            
            if(!m[s[i]]){
                m[s[i]]++;
            }

            res = max(res,i - l + 1);
        }

        return res;
    }
};

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
    const set = new Set();
    let l = 0;
    let res = 0;

    for(let i = 0; i<s.length; i++){
        let st = s.charAt(i);
        while(set.has(st) && l <= i){
            set.delete(s.charAt(l++));
        }
        set.add(st);
        res = Math.max(res,i - l + 1);
    }
    return res;
};